src/bp/core/services/telemetry/sdk-methods.ts

```diff
@@ -0,0 +1,185 @@
+import { parse } from '@babel/parser'
+import traverse from '@babel/traverse'
+import LicensingService from 'common/licensing-service'
+import { buildSchema } from 'common/telemetry'
+import Database from 'core/database'
+import { TelemetryRepository } from 'core/repositories/telemetry'
+import { TYPES } from 'core/types'
+import { inject, injectable } from 'inversify'
+import _ from 'lodash'
+import ms from 'ms'
+
+import { GhostService } from '..'
+import { BotService } from '../bot-service'
+import { JobService } from '../job-service'
+
+import { BotHooks, getHooksLifecycle, Hook } from './hooks'
+import { TelemetryStats } from './telemetry-stats'
+
+const PARSE_CONFIG = { allowReturnOutsideFunction: true }
+
+interface UsageParams {
+  botId?: string
+  lifecycle?: string
+  enabled?: Boolean
+}
+
+interface Usage {
+  module: string
+  function: string
+  count: number
+}
+
+interface ParsedFile extends UsageParams {
+  fileName: string
+  usages: Usage[]
+}
+
+interface SDKUsageEvent {
+  actions: SDKUsage
+  hooks: SDKUsage
+}
+
+interface SDKUsage {
+  global: ParsedFile[]
+  perBots: ParsedFile[]
+}
+
+@injectable()
+export class SDKStats extends TelemetryStats {
+  protected interval: number
+  protected url: string
+  protected lock: string
+
+  constructor(
+    @inject(TYPES.GhostService) ghostService: GhostService,
+    @inject(TYPES.Database) database: Database,
+    @inject(TYPES.LicensingService) licenseService: LicensingService,
+    @inject(TYPES.JobService) jobService: JobService,
+    @inject(TYPES.TelemetryRepository) telemetryRepo: TelemetryRepository,
+    @inject(TYPES.BotService) private botService: BotService
+  ) {
+    super(ghostService, database, licenseService, jobService, telemetryRepo)
+    this.url = process.TELEMETRY_URL
+    this.lock = 'botpress:telemetry-SDK'
+    this.interval = ms('1d')
+  }
+
+  protected async getStats() {
+    return {
+      ...buildSchema(await this.getServerStats(), 'server'),
+      event_type: 'custom_roles',
+      event_data: { schema: '1.0.0', SDKMethods: await this.SDKUsage() }
+    }
+  }
+
+  private async SDKUsage(): Promise<SDKUsageEvent> {
+    const bots = await this.botService.getBotsIds()
+    return { actions: await this.getActionUsages(bots), hooks: await this.getHooksUsages() }
+  }
+
+  private async getActionUsages(bots: string[]): Promise<SDKUsage> {
+    const rootFolder = 'actions'
+    const globalActionsNames = await this.ghostService
+      .global()
+      .directoryListing('/', `${rootFolder}/*.js`)
+      .map(path => path.split('/').pop() || '')
+
+    const global = await Promise.map(globalActionsNames, this.parseFile(`/${rootFolder}`))
+    const perBots = await Promise.reduce(bots, this.botActionsReducer(rootFolder), [] as ParsedFile[])
+
+    return { global, perBots }
+  }
+
+  private botActionsReducer(rootFolder: string): (acc: any, botId: any) => Promise<ParsedFile[]> {
```

Comment:
why any types?

```
private botActionsReducer(rootFolder: string): (acc: ParsedFile[], botId: string) => Promise<ParsedFile[]> {
...
}
```

\+ I'm guilty of this too, but naming the accumulator `acc` is as lazy as naming a variable `vari` or a parameter: `param`. Comon le gros.

Reply:
fixed:  64d2839d0b1bfe78162d88070543b062b7de4b9a and  17e2c076b0bb16928d097815ce71a66b2268c959